src/runtime-rs/crates/hypervisor/src/ch/inner_device.rs

```diff
@@ -25,17 +25,44 @@ const DEFAULT_DISK_QUEUE_SIZE: u16 = 1024;
 impl CloudHypervisorInner {
     pub(crate) async fn add_device(&mut self, device: DeviceType) -> Result<()> {
         if self.state != VmmState::VmRunning {
-            let mut devices: Vec<DeviceType> = if let Some(devices) = self.pending_devices.take() {
-                devices
+            // If the VM is not running, add the device to the pending list to
+            // be handled later.
+            //
+            // Note that the only device type considered is DeviceType::ShareFs since:
+            //
+            // - ShareFs (virtiofsd) is only needed in an non-DM and non-TDX scenario
+            //   for the container rootfs.
+            //
+            // - For all other scenarios, the container rootfs is handled by a
+            //   DeviceType::Block and this method is called *after* the VM
+            //   has started so the device does not need to be added to the
+            //   pending list.
+            //
+            // - The VM rootfs is handled without waiting for calls to this
+            //   method as the file in question (image= or initrd=) is available
+            //   from HypervisorConfig.BootInfo.{image,initrd}
+            //   (see 'convert.rs').
+            //
+            if let DeviceType::ShareFs(_) = device {
```

Comment:
Network device is also considered a sandbox resource and needs to be added before the VM starts running. This change may break that case?

Reply:
Fixed.